src/serializer.ts

```diff
@@ -0,0 +1,187 @@
+/**
+ * ObjectTemplate.toJSONString;
+ */
+/**
+ * Convert an object to JSON, stripping any recursive object references so they can be
+ * reconstituted later
+ *
+ * @param {unknown} obj unknown
+ * @param {unknown} cb unknown
+ *
+ * @returns {unknown}
+ */
+export function toJSONString(obj, cb?) {
+    var idMap = [];
+
+    try {
+        return JSON.stringify(obj, function a(key, value) {
+            if (key === '__objectTemplate__' || key === 'amorphic') {
+                return null;
+            }
+            if (value && value.__template__ && value.__id__) {
+                if (idMap[value.__id__]) {
+                    value = { __id__: value.__id__.toString() };
+                }
+                else {
+                    idMap[value.__id__.toString()] = value;
+                }
+            }
+
+            if (cb) {
+                return cb(key, value);
+            }
+
+            return value;
+        });
+    }
+    catch (e) {
+        throw e;
+    }
+}
+
+function propXfer(prop, pojo, obj) {
+    if (pojo[prop]) {
+        obj[prop] = pojo[prop];
+    }
+}
+
+/**
+ * ObjectTemplate.fromPOJO
+ */
+export function fromPOJO(pojo, template, defineProperty?, idMap?, idQualifier?, parent?, prop?, creator?) {
+    function getId(id) {
+        if (typeof (idQualifier) !== 'undefined') {
+            return id + '-' + idQualifier;
+        }
+
+        return id;
+    }
+
+    // For recording back refs
+    if (!idMap) {
+        idMap = {};
+    }
+
+    if (!pojo.__id__) {
+        return;
+    }
+
+    var obj;
+
+    if (creator) {
+        obj = creator(parent, prop, template, idMap[pojo.__id__.toString()], pojo.__transient__);
+
+        if (obj instanceof Array) {
+            obj = obj[0];
+            idMap[obj.__id__.toString()] = obj;
+            return obj;
+        }
+
+        if (typeof (obj) === 'undefined') {
+            return null;
+        }
+
+        if (!obj) {
+            this.noInit = true;
+            obj = new template();
+            this.noInit = false;
+        }
+    }
+    else {
+        obj = this._createEmptyObject(template, getId(pojo.__id__.toString()), defineProperty, pojo.__transient__);
+    }
+
+    idMap[getId(pojo.__id__.toString())] = obj;
+
+    // Go through all the properties and transfer them to newly created object
+    var props = obj.__template__.getProperties();
+
+    for (var propb in pojo) {
+        propb = propb.replace(/^__/, '');
+        var defineProp = props[propb];
+        if (!defineProp)
+            continue;
+        var type = defineProp.type;
+
+        // Because semotus can serialize only the shadow properties we try and restore them
+        var pojoProp = (type && typeof pojo['__' + propb] !== 'undefined') ? '__' + propb : propb;
```

Comment:
May be a "Todo" item for future, if semotus needs these shadow properties to be restored, the functionality should be moved to semotus. We should clearly segregate the responsibilities of these libraries.